### DISCUSSION TITLE
How to type a class decorator registry pattern?

### DISCUSSION THREAD

**Discussion by cdce8p** [python/typing#1565] [Q&A]
Came across a pattern today that left me wondering how it **should** be typed instead or if it's even possible to type probably.
It works for Mypy but I'm guessing this is just because class decorators aren't type checked at all https://github.com/python/mypy/issues/3135.

```py
from typing import TypeVar, Callable

_KT = TypeVar("_KT")
_VT = TypeVar("_VT")

class Registry(dict[_KT, _VT]):
    def register(self, name: _KT) -> Callable[[_VT], _VT]:
        def decorator(func: _VT) -> _VT:
            self[name] = func
            return func
        return decorator


class Provider: ...

PROVIDERS: Registry[str, type[Provider]] = Registry()

@PROVIDERS.register("auth")
class AuthProvider(Provider): ...
```

```py
def a1() -> None:
    reveal_type(AuthProvider)              # this should be type[AuthProvider]
    reveal_type(list(PROVIDERS.values()))  # this should be list[type[Provider]]

    @PROVIDERS.register("error")           # this should be an error
    class Error: ...
```

**Comment by Daverball** [comment]
`_VT` should probably have an upper bound of `Callable[..., Any]`, since `register` will only work for functions or classes. Or if you actually want to restrict it to types you could also do `class Registry(dict[_KT, type[_VT]]):` and replace all the other occurrences of `_VT` with `type[_VT]`.

Class decorators are type checked now, but they can't change the type, which should be fine for your use-case. It looks like your example code already works just fine with the behavior you expect: https://mypy-play.net/?mypy=latest&python=3.12&gist=9843acb42af5acef8f7a982d2c7207ac

**Reply by cdce8p** [reply]
> `_VT` should probably have an upper bound of `Callable[..., Any]`, since `register` will only work for functions or classes. Or if you actually want to restrict it to types you could also do `class Registry(dict[_KT, type[_VT]]):` and replace all the other occurrences of `_VT` with `type[_VT]`.

👍🏻 Good point. It's actually included in the original source code, just forgot to copy it to the example.

> Class decorators are type checked now, but they can't change the type, which should be fine for your use-case. It looks like your example code already works just fine with the behavior you expect: https://mypy-play.net/?mypy=latest&python=3.12&gist=9843acb42af5acef8f7a982d2c7207ac

🤔 Does that apply for all type checkers that class decorators should not be able to change the type?
It seems that's how both `Mypy` and `Pyre` interpret it that way. However the example doesn't work with `pyright`.

[pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAMLkA2D5ARg8QFAcD6A0gVAC8hEhRAAKAER8CkgJQ9SA4UQRlKU7kvlcAxswDOBqACViaJAZgg44gCZJdMANoyaWggF05ALg5QAqDtiYCgQc0sYYgkDYgZgGhRyCGIfKBk5KABaAD46RmY2YmdXJU93Mr9A6qCQ2t1cchhccWAAVxRdNI9M3PSlKpqh2PjnJJTPISh2zv8hmvCYNpAUaY7dOaHF5dXghpAm3C4OfXIjKAAFcAA3JGCQNIA6Z64LkwB5UgBJABEAURMAGU0mYLFYbM5wTR4GpnFcwLd7p5JsJQZEbOIFBwAAJvT6-AGAx7hMFRCSSchtGAACx0p3OAEEqdT4YjouJWXdor4oM9HlxgqFyABGTHZPIAOTAKFSmzCxGuxEY3BhxHETJpnPumXmQwAxLBqZYoAZqWA2gw7FAWMRYCRnBqWTcuSBPHLworlarxAxIhyPt9-kDHtdGG1iAZMXIdQaacbTebLdbbb6rM5VXDnUi3XLcQGCcGSZF2ZJouAQPJdYFY0bjAmLVabVByKsy0dqvTjH8QOWns8gA)
[pyre playground](https://pyre-check.org/play?input=from%20typing%20import%20TypeVar%2C%20Callable%0A%0A_KT%20%3D%20TypeVar(%22_KT%22)%0A_VT%20%3D%20TypeVar(%22_VT%22)%0A%0Aclass%20Registry(dict%5B_KT%2C%20_VT%5D)%3A%0A%20%20%20%20def%20register(self%2C%20name%3A%20_KT)%20-%3E%20Callable%5B%5B_VT%5D%2C%20_VT%5D%3A%0A%20%20%20%20%20%20%20%20def%20decorator(func%3A%20_VT)%20-%3E%20_VT%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20self%5Bname%5D%20%3D%20func%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20func%0A%20%20%20%20%20%20%20%20return%20decorator%0A%0A%0Aclass%20Provider%3A%20...%0A%0APROVIDERS%3A%20Registry%5Bstr%2C%20type%5BProvider%5D%5D%20%3D%20Registry()%0A%0A%40PROVIDERS.register(%22auth%22)%0Aclass%20AuthProvider(Provider)%3A%20...%0A%0Adef%20a1()%20-%3E%20None%3A%0A%20%20%20%20reveal_type(AuthProvider)%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20this%20should%20be%20type%5BAuthProvider%5D%0A%20%20%20%20reveal_type(list(PROVIDERS.values()))%20%20%23%20this%20should%20be%20list%5Btype%5BProvider%5D%5D%0A%0A%20%20%20%20%40PROVIDERS.register(%22error%22)%20%20%20%20%20%20%20%20%20%20%20%23%20this%20should%20be%20an%20error%0A%20%20%20%20class%20Error%3A%20...)

/CC: @erictraut

**Reply by Daverball** [reply]
The pyright output looks correct to me as well, maybe the error message is what is confusing you?

**Reply by cdce8p** [reply]
> The pyright output looks correct to me as well, maybe the error message is what is confusing you?

Whereas both Mypy and pyre infer `reveal_type(AuthProvider)` as `type[AuthProvider]`, pyright uses the decorator typing and infers `type[Provider]`. This causes issues for me down the line where `AuthProvider` is expected, as even with `var: AuthProvider`,  `var` would only be inferred as `Provider`, and thus missing the methods that only `AuthProvider` has.

That's why I was asking if my initial typing was wrong.

**Reply by Daverball** [reply]
Oh, I see. Yeah I missed that and I also didn't really think hard enough about that part.

What you really want is for the decorator to remain generic, regardless of what you bind the registry to (binding the type var would only change the upper bound). I unfortunately don't think there's a way to express that, it would require some higher kinded `TypeVar` which could set its upper bound to the value of another `TypeVar`.

For now I would probably just either create separate non-generic `Registry` subclasses for each type, where the method itself can remain generic, or give up some amount of type safety and use a different `TypeVar` in the `register` method.

Maybe in some hypothetical future of Python we'll be allowed to do things like this with PEP695 syntax, since there's not as much ambiguity about scoping, to allow doing things like changing the upper bound based on another `TypeVar` in the scope of a generic method within a generic class.

```python
class Registry[_KT, _VT](dict[_KT, _VT]):
    def register[_T: _VT](self, name: _KT) -> Callable[[_T], _T]:
        def decorator(func: _T) -> _T:
            self[name] = func
            return func
        return decorator
```

**Reply by Daverball** [reply]
In a future with intersection this might be another potential solution
```python
def register[_T](self, name: _KT) -> Callable[[_T & _VT], _T]:
```
Although the semantics here are a bit more questionable, since it's unclear what `_T` should bind to when it's part of an intersection.

**Comment by Akuli** [comment]
You could cast the decorator to `_typeshed.IdentityFunction` so it will always return the type passed into it:

```python
from _typeshed import IdentityFunction

...

class Registry(dict[_KT, _VT]):
    def register(self, name: _KT) -> IdentityFunction:
        def decorator(func: _VT) -> _VT:
            self[name] = func
            return func
        return cast(IdentityFunction, decorator)
```

[Pyright playground link](https://pyright-play.net/?code=GYJw9gtgBA%2BgLgTwA4FMDOALFATKBLCJMEOKASWxQDs49EAxAVyoGNawqAoT0SKRJHioBzfIWKkAKshQA1AIYgANFBby0cbjADSkqAF4o01ApAAKAEQ7JFgJScYsvYeNzFlxzfucWAG3VoUABKKMJ4GiAIZth4bADa1iqeALq2AFycUFlQlMBQIKHhcCjmaCi%2BwCpU8hAoabC6tlAAtAB85JQ0dAhMrOxUGdlDOSh5lCzE8nDEZsDMLPWeTW2wToPDG2UVcdW1yQZQc6yZG8MFcIwgVIfzJxvnl9dqGmYU1LQM8-0q45PTIN4fP40IEAArgABueEoIHqADoEdxQUEAPKyMgAEQAokEAMr1EJhCIIOIRFQCFBxcFgKEw5L7QyEoqRMzeAACyLRmJxuLhBSJxXMFnkjDgGDsQICUAAgqKMNTaSUzAroSV0lAEXDOLkoPIAIyslrtAByHDqd3yKAhKHkvngMjMsrFKphTVOGwAxPwMOEoJgwIxfLgAEYofgyOJO%2BWQ1UgZIWgrW2321BmXxFZWo9HYvFwiG2xjoVm2N1esW%2B-2BkNh9MaOIUqkxunxi0crPc3P8opKiwlcAgOzu7Jln2BStBqCh3XXPvEC1%2BKVYkD9%2BGIoA)

**Reply by cdce8p** [reply]
`IdentifyFunction` is a nice trick but, unless I'm missing something, I'd still lose the error detection for this unfortunately
```py
    @PROVIDERS.register("error")           # this should be an error
    class Error: ...
```

**Reply by Daverball** [reply]
Yeah, pretty much, you can't have both right now if you want to support pyright (at least if you go with the generic `Registry`, for a specific `Registry` you can achieve it using type var upper bounds, but that also means you need a new type var and `Registry` class definition for each type of registry).

And even if you didn't care about supporting pyright, there's a chance that mypy and other type checkers will respect the return type of the decorator in the future and then it'll break anyways.
